### ISSUE TITLE
Can't set default values for an array of ParameterType.COMPLEX objects due to type error

### ISSUE THREAD

**Issue by MichaelGoodale** [jspsych/jsPsych#2882]
Hello, 

I'm developing a little plugin and I would like to set a default value for a parameter which is an array of objects (really I'd prefer to just have an array of arrays of strings, but this seems the better way to do it in jsPsych). While it works brilliantly as long as I provide that parameter, I can't seem to provide a default value without throwing a type error.  I've tried various options for the default value: `null`, `[]`, or a fully specified array: `[{members_ids: ['a']}]` but unfortunately they all throw the same error.

Ultimately, I can avoid the problem by providing a default to the parameter each time I use the plugin, but it's slightly annoying to have to always fill in that parameter, and I assume this is not the intended behaviour.  I'm by no means a JS expert so perhaps there is something basic I'm missing, but I couldn't find anything that would fix the problem.

### Error
```
Uncaught (in promise) TypeError: t[i] is undefined
    setDefaultValues https://unpkg.com/jspsych@7.0.0:1
    doTrial https://unpkg.com/jspsych@7.0.0:1
    startExperiment https://unpkg.com/jspsych@7.0.0:1
    jsPsychModule https://unpkg.com/jspsych@7.0.0:1
    o https://unpkg.com/jspsych@7.0.0:1
    promise callback*l https://unpkg.com/jspsych@7.0.0:1
    o https://unpkg.com/jspsych@7.0.0:1
    promise callback*l https://unpkg.com/jspsych@7.0.0:1
    o https://unpkg.com/jspsych@7.0.0:1
    promise callback*l https://unpkg.com/jspsych@7.0.0:1
    jsPsychModule https://unpkg.com/jspsych@7.0.0:1
    n https://unpkg.com/jspsych@7.0.0:1
    run https://unpkg.com/jspsych@7.0.0:1
    <anonymous> file:///home/michael/Documents/plural-causes-experiment/example.html:48
jspsych@7.0.0:1:37357
    o https://unpkg.com/jspsych@7.0.0:1
    (Async: promise callback)
    l https://unpkg.com/jspsych@7.0.0:1
    o https://unpkg.com/jspsych@7.0.0:1
    (Async: promise callback)
    l https://unpkg.com/jspsych@7.0.0:1
    o https://unpkg.com/jspsych@7.0.0:1
    (Async: promise callback)
    l https://unpkg.com/jspsych@7.0.0:1
    jsPsychModule https://unpkg.com/jspsych@7.0.0:1
    n https://unpkg.com/jspsych@7.0.0:1
    run https://unpkg.com/jspsych@7.0.0:1
    run self-hosted:1356
    <anonymous> file:///home/michael/Documents/plural-causes-experiment/example.html:48
```


### Minimal code to reproduce bug
```
<!DOCTYPE html>
<html>
  <head>
    <title>My experiment</title>
    <script src="https://unpkg.com/jspsych@7.0.0"></script>
    <script src="https://unpkg.com/@jspsych/plugin-html-keyboard-response@1.0.0"></script>
    <link href="https://unpkg.com/jspsych@7.0.0/css/jspsych.css" rel="stylesheet" type="text/css" />
  </head>
<body></body>
<script>
  var jsPluginName = (function (jspsych) {
    "use strict";
    const info = {
      name: "PLUGIN-NAME",
      parameters: {
        /** List of groups of arrays */
        groups: {
          type: jspsych.ParameterType.COMPLEX,
          array: true,
          default: null,
          nested: {
            /** List of members of groups */
            members_id: {
              type: jspsych.ParameterType.STRING,
              pretty_name: "Group members' ID",
              array: true,
              default: null,
            }
          }
        },
  
      },
    };
    class PluginNamePlugin {
      constructor(jsPsych) {
        this.jsPsych = jsPsych;
      }
      trial(display_element, trial) {
        this.jsPsych.finishTrial();
      }
    }
    PluginNamePlugin.info = info;
  
    return PluginNamePlugin;
  })(jsPsychModule);

  var jsPsych = initJsPsych();
  jsPsych.run([{type: jsPluginName}]);
  </script>
</html>
```

**Comment by nikbpetrov** [comment]
@MichaelGoodale Great spot! Looking at a similar plugin behaviour, the [survey-text](https://www.jspsych.org/7.3/plugins/survey-text/), which has a nested parameter, namely `questions`, throws the same error.

It turns out that the problem lies in the `setDefaultValues` function, which, when it checks for default parameters, has the logic backwards (or at least very weirdly): it goes through all of your definitions of different nested stuff (e.g. your many groups in your case) and for each one it checks whether it has a specific nested parameter defined - if not, it sets it. In other words, if you have this:

```javascript
parameters: {
        groups: {
          type: jspsych.ParameterType.COMPLEX,
          array: true,
          pretty_name: "Groups",
          default: undefined,
          nested: {
            members_id: {
              type: jspsych.ParameterType.STRING,
              pretty_name: "Members ids",
              array: true,
              default: "id1",
            }
          }
        }

var trial = {
    type: jsPluginName,
    groups: [{}, {}]
  }
```

would be the same as this:

```javascript
parameters: {
        groups: {
          type: jspsych.ParameterType.COMPLEX,
          array: true,
          pretty_name: "Groups",
          default: undefined,
          nested: {
            members_id: {
              type: jspsych.ParameterType.STRING,
              pretty_name: "Members ids",
              array: true,
              default: undefined,
            }
          }
        }

var trial = {
    type: jsPluginName,
    groups: [{members_id: "id1"}, {members_id: "id1"}]
  }
```

Crucially, however, at no point does jsPsych process the `array: true` part that is placed INSIDE the nested parameter - in this case the `members_id`. So if you wanted to have, by default, two groups with ids "id1" and "id2", this cannot be done, let alone further recursive nested (i.e. having `members_id` as `jspsych.ParameterType.COMPLEX`

I will put together a hotfix for your case shortly.

Tagging @jodeleeuw and @bjoluc , who are working on v8 codebase and might want to check if that is addressed.

**Comment by bjoluc** [comment]
Thanks @nikbpetrov and @MichaelGoodale! This works in the v8 draft, i.e.

https://github.com/jspsych/jsPsych/blob/bbdf1585645874277ade36c7de433064d417aba6/packages/jspsych/src/timeline/Trial.spec.ts#L298-L305

with input

https://github.com/jspsych/jsPsych/blob/bbdf1585645874277ade36c7de433064d417aba6/packages/jspsych/src/timeline/Trial.spec.ts#L326-L331

yields

https://github.com/jspsych/jsPsych/blob/bbdf1585645874277ade36c7de433064d417aba6/packages/jspsych/src/timeline/Trial.spec.ts#L347-L350

I would appreciate if you could use @nikbpetrov's hotfix for now @MichaelGoodale :pray:

**Comment by MichaelGoodale** [comment]
Thanks so much for the quick response @nikbpetrov and @bjoluc, this will be very helpful!

**Comment by jodeleeuw** [comment]
I just fixed this in #2974 for version 7.3.2.
